### ISSUE TITLE
Set up instructions (Ubuntu)

### ISSUE THREAD

**Issue by dominklenda** [ScienceCommons/curate_science#115]
Detailed and reorganized instructions based on READ.md and issue #76 prepared on Ubuntu, POP!_OS 19.10.

### Prerequisites
- Python >= 3.6
- npm

### Update OS and clone this repo
```bash
sudo apt update
sudo apt upgrade
git clone https://github.com/ScienceCommons/curate_science.git
```

### Python set up
__Check if pip3 and python3 are installed__
```bash
pip3 -V
python3 -V
```
__Create and activate virtual environment__
```bash
cd /home/user_name/your_path/curate_science
python3 -m venv cs_env
source cs_env/bin/activate
```
__Install all necessary Python packages__
Make sure that ```wheel``` is installed ```wheel-V```, if not, install it using ```pip3 install wheel```. Then install all required Python packages, e.g. Django, Google Cloud Storage. 
```bash
pip3 install -r requirements.txt
```

### PostgreSQL set up
__Install PostgreSQL  >= 9.4__
```sql
sudo apt install postgresql postgresql-contrib
```

__Create .env file and set environmental variables with your own name, password, and key__
```sql
DB_USER=user_name
DB_PASS=user_password
SECRET_KEY=key
```
Login to postgresql interactive terminal
```bash
sudo -u postgres psql
```
__Create a database, user, and add access__
```sql
CREATE DATABASE curate;
CREATE USER [DB_USER] WITH PASSWORD '[DB_PASS]';
GRANT ALL PRIVILEGES ON DATABASE curate TO [DB_USER];
ALTER USER [DB_USER] CREATEDB;
```

__Get backup data__
*Email Curate Science [admin](admin@curatescience.org) to get access to the current backup file*

Quit PostgreSQL shell and restore dump. Here its name is ```backup.bak```
```bash
\q
psql curate < backup.bak
```

### Django set up
__Execute Django migrations__
Make new migrations
```bash
python manage.py makemigrations
```
Apply migrations
```bash
python3 manage.py migrate
```

__Access Django admin__
Fill in your username, email, password
```bash
python3 manage.py createsuperuser
```

### Get images
See [Google Cloud API](https://cloud.google.com/storage/docs/gsutil_install#deb) to install `gsutil`

__Copy files from the staging bucket__
Create a new folder `media` that contains images and copy files from the bucket using multithreaded `gsutil cp` 
```bash
mkdir media
cd media
gsutil -m cp -r gs://curate-science-staging-2.appspot.com/key_figures .
```

### Preparation for development
Check if `npm` is installed. Install dependencies, `yarn` globally and `webpack` locally
```bash
npm -v
npm install
npm install -g yarn
npm install webpack
```

### Running Tests
```
python3 manage.py test
```

### Starting local server for development
```bash

python3 manage.py runserver_plus
# In a new terminal (to compile JS bundle(s) and HTML)
yarn start
```

App should be running at `localhost:8000`
Django admin should be available at `localhost:8000/admin`

**Comment by eplebel** [comment]
great, thanks so much for this!
i've now added a link to your more detailed dev setup instructions within the README file, see https://github.com/ScienceCommons/curate_science/blob/master/README.md#setup

**Comment by dominklenda** [comment]
Today I updated my OS to Ubuntu 20.04 and the app did not work. I tried to restore the virtual environment, but couldn't run `python manage.py runserver_plus` afterward. I had to change the version of psycopg in the requirements.txt file before the installation of python dependencies 
from 
`psycopg2==2.7.5
psycopg2-binary==2.7.5`
to
`psycopg2==2.8
psycopg2-binary==2.8
`
Somehow psycopg 2.7.5 couldn't be installed using pip. Now everything's fine.
